caffe2/operators/batch_bucketize_op.cc

```diff
@@ -0,0 +1,126 @@
+#include "batch_bucketize_op.h"
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/tensor.h"
+
+namespace caffe2 {
+
+template <>
+bool BatchBucketizeOp<CPUContext>::RunOnDevice() {
+  auto& indices = Input(INDICES);
+  auto& borders = Input(BORDERS);
+  auto& lengths = Input(LENGTHS);
+  auto& feature = Input(FEATURE);
+  auto* output = Output(O);
+  CAFFE_ENFORCE_EQ(lengths.ndim(), 1);
+  CAFFE_ENFORCE_EQ(indices.ndim(), 1);
+  CAFFE_ENFORCE_EQ(borders.ndim(), 1);
+  CAFFE_ENFORCE_EQ(feature.ndim(), 2);
+  CAFFE_ENFORCE_EQ(lengths.size(), indices.size());
+
+  const auto* lengths_data = lengths.template data<int32_t>();
+  const auto* indices_data = indices.template data<int32_t>();
+  const auto* borders_data = borders.template data<float>();
+  const auto* feature_data = feature.template data<float>();
+  auto batch_size = feature.dim(0);
+  auto feature_dim = feature.dim(1);
+  auto output_dim = indices.size();
+
+  TIndex length_sum = 0;
+  for (TIndex i = 0; i < lengths.size(); i++) {
+    CAFFE_ENFORCE_GE(feature_dim, indices_data[i]);
+    length_sum += lengths_data[i];
+  }
+  CAFFE_ENFORCE_EQ(length_sum, borders.size());
+
+  TIndex lower_bound = 0;
+  output->Resize(batch_size, output_dim);
+  auto* output_data = output->template mutable_data<int32_t>();
+
+  for (TIndex i = 0; i < batch_size; i++) {
+    lower_bound = 0;
+    for (TIndex j = 0; j < output_dim; j++) {
+      for (TIndex k = 0; k <= lengths_data[j]; k++) {
+        if (k == lengths_data[j] ||
+            feature_data[i * feature_dim + indices_data[j]] <
+                borders_data[lower_bound + k]) {
+          output_data[i * output_dim + j] = k;
+          break;
+        } else {
+          continue;
+        }
+      }
+      lower_bound += lengths_data[j];
+    }
+  }
+  return true;
+}
+
+REGISTER_CPU_OPERATOR(BatchBucketize, BatchBucketizeOp<CPUContext>);
+
+OPERATOR_SCHEMA(BatchBucketize)
+    .NumInputs(4)
+    .NumOutputs(1)
+    .SetDoc(R"DOC(
+Bucketize the float_features into sparse features.
+The float_features is a N * D tensor where N is the batch_size, and D is the feature_dim.
+The indices is a 1D tensor containing the indices of the features that need to be bucketized.
```

Comment:
I think you can first do Gather/BatchGather instead of adding this `indices` argument into `BatchBucketize`

Reply:
Will change that.

Reply:
I think it'll be better to bucketize features by one operator. Maybe we can set up a meeting tomorrow to talk about this.

Reply:
We in general discourage fused operators (unless it's for performance reason in critical operators, like `bias` in Convolution), because it's in general less expressive.

Reply:
@bddppq The extra code here for a fused operation is minimal, and it doesn't make the operator any less expressive. On the other hand, When you're talking about large amounts of data flowing through for preprocessing features, it reduces copying the data by 2, most likely making the operator almost 2 times faster. (as opposed to using BatchGather first and then this operator)

Reply:
@amirshim In your use case, normally what's the size of the batched features matrix?

Reply:
chatted offline, fusion is a right compromise here.